packages/spring/test/index.test.ts

```diff
@@ -0,0 +1,164 @@
+import { describe, expect, it } from "vitest";
+import { createRoot, createSignal } from "solid-js";
+import { createDerivedSpring, createSpring } from "../src/index.js";
+
+// ===========================================================================
+// Tests
+// ===========================================================================
+
+describe("createSpring", () => {
+  it("returns values", () => {
+    const { value, setValue, dispose } = createRoot(dispose => {
+      const [value, setValue] = createSpring({ progress: 0 });
+      expect(value().progress, "initial value should be { progress: 0 }").toBe(0);
+
+      return { value, setValue, dispose };
+    });
+
+    setValue({ progress: 100 });
+    dispose();
+  });
+
+  it("updates toward target", async () => {
+    let dispose!: () => void;
+    let setSpringed!: (
+      newValue: number,
+      opts?: { hard?: boolean; soft?: boolean },
+    ) => Promise<void>;
+    let springed!: () => number;
+
+    createRoot(d => {
+      dispose = d;
+      [springed, setSpringed] = createSpring(0);
+    });
+
+    expect(springed()).toBe(0);
+    setSpringed(50); // Set to 100 here.
+
+    // Not sure if this will be erratic.
+    await new Promise(resolve => setTimeout(resolve, 300));
+
```

Comment:
Yeah timeouts in tests are my worst nightmare. If you can mock the animation frame and test it without using timeout it would be much better.
The tween package does that so possibly it could be copied: https://github.com/solidjs-community/solid-primitives/blob/main/packages/tween/test/index.test.ts

Reply:
Unfortunate news, I can't mock it. I tried:

1. vi.useFakeTimers and vi.advanceTimersByTime
```ts
describe("createSpring", () => {
  it("updates toward target over time", async () => {
    vi.useFakeTimers(); // Use fake timers

    const { dispose, springed, setSpringed } = createRoot(dispose => {
      const [springed, setSpringed] = createSpring(0);
      return { dispose, springed, setSpringed };
    });

    expect(springed()).toBe(0);
    setSpringed(100); // Set target to 100

    // Advance timers and check intermediate states
    vi.advanceTimersByTime(100);
    expect(springed()).toBeGreaterThan(0);
    expect(springed()).toBeLessThan(100);

    vi.advanceTimersByTime(500);
    expect(springed()).toBeGreaterThan(50);
    expect(springed()).toBeLessThan(100);

    vi.advanceTimersByTime(1000);
    expect(springed()).toBe(100); // Should have reached the target

    dispose();
  });
});
``` 

2. Complete copy of the `_flush_raf` function in `tween`
```ts
let _last_id = 0;
let _raf_callbacks_old = new Map<number, FrameRequestCallback>();
let _raf_callbacks_new = new Map<number, FrameRequestCallback>();

function _flush_raf(time: number) {
  _raf_callbacks_old = _raf_callbacks_new;
  _raf_callbacks_new = new Map();
  for (const callback of _raf_callbacks_old.values()) {
    callback(time);
  }
  _raf_callbacks_old.clear();
}

function _mocked_requestAnimationFrame(callback: FrameRequestCallback): number {
  const id = _last_id++;
  _raf_callbacks_new.set(id, callback);
  return id;
}
function _mocked_cancelAnimationFrame(id: number): void {
  _raf_callbacks_new.delete(id);
}

vi.stubGlobal("requestAnimationFrame", _mocked_requestAnimationFrame);
vi.stubGlobal("cancelAnimationFrame", _mocked_cancelAnimationFrame);

afterEach(() => {
  _raf_callbacks_old.clear();
  _raf_callbacks_new.clear();
  _last_id = 0;
});

describe("createSpring", () => {
  it("updates toward target over time", async () => {
    vi.useFakeTimers(); // Use fake timers

    const { dispose, springed, setSpringed } = createRoot(dispose => {
      const [springed, setSpringed] = createSpring(0);
      return { dispose, springed, setSpringed };
    });

    expect(springed()).toBe(0);
    setSpringed(100); // Set target to 100

    // Advance timers and check intermediate states
    _flush_raf(start + 25);
    expect(springed()).toBeGreaterThan(0);
    expect(springed()).toBeLessThan(100);

    _flush_raf(start + 50);
    expect(springed()).toBeGreaterThan(50);
    expect(springed()).toBeLessThan(100);

    _flush_raf(start + 100);
    expect(springed()).toBe(100); // Should have reached the target

    dispose();
  });
});

```

3. Exported `raf` from the implementation and overriding it.
4. Mocking `raf`

To no avail. In any case, I'll just refactor the tests to the best I can for now (based on your revisions). Thanks again!